src/dot-protocol/merlin_dot_protocol.mli

```diff
@@ -74,19 +74,53 @@ end
 
 type directive = Directive.Processed.t
 
-module Commands : sig
-  type t = File of string | Halt | Unknown
+module type S = sig
+  type 'a io
+  type in_chan
+  type out_chan
 
-  val read_input : in_channel -> t
-  val send_file : out_channel:out_channel -> string -> unit
-end
-
-type read_error =
+  type read_error =
   | Unexpected_output of string
   | Csexp_parse_error of string
 
-(** [read inc] reads one csexp from the channel [inc] and returns the list of
-  directives it represents *)
-val read : in_channel:in_channel -> (directive list, read_error) Merlin_utils.Std.Result.t
+  (** [read] reads one csexp from the channel and returns the list of
+      directives it represents *)
+  val read :
+    in_chan -> (directive list, read_error) Merlin_utils.Std.Result.t io
+
+  val write : out_chan -> directive list -> unit io
+
+  module Commands : sig
```

Comment:
If a type doesn't depend on any of the functor arguments, it would be cleaner to move it out and make it usable in a wider context. Same for `read_error`.

I think this is how I did in my original patch.